mypy/util.py

```diff
@@ -183,24 +188,39 @@ def correct_relative_import(cur_mod_id: str,
     return cur_mod_id + (("." + target) if target else ""), ok
 
 
+FIELDS_CACHE = {}  # type: Dict[Type[object], List[str]]
+
+
 def replace_object_state(new: object, old: object) -> None:
     """Copy state of old node to the new node.
 
-    This handles cases where there is __slots__ and/or __dict__.
+    This handles cases where there is __dict__ and/or attribute descriptors
+    (either from slots or because the type is defined in a C extension module).
 
     Assume that both objects have the same __class__.
     """
     if hasattr(old, '__dict__'):
         new.__dict__ = old.__dict__
-    if hasattr(old, '__slots__'):
-        # Use __mro__ since some classes override 'mro' with something different.
-        for base in type(old).__mro__:
-            if '__slots__' in base.__dict__:
-                for attr in getattr(base, '__slots__'):
-                    if hasattr(old, attr):
-                        setattr(new, attr, getattr(old, attr))
-                    elif hasattr(new, attr):
-                        delattr(new, attr)
+
+    cls = old.__class__
+    # Maintain a cache of type -> attributes
```

Comment:
This doesn't include attributes in `__dict__`. It would be better to explicitly enumerate which attributes are processed here (attributes in `__slots__` and attributes defined in C extension classes?).